# Анализ трех тикеров

df_transfap  - первый тикет

df_krasoctao - второй тикет

df_irkut3  - третий тикет

### Загрузка данных

In [34]:
import pandas as pd #библиотека для визуализации данных
from IPython.display import display #функция позволит явно вывести объект
import numpy as np 

df_transfap = pd.read_csv("transfap") #первый тикет
df_krasoctao = pd.read_csv("КрасОкт_ао") #второй тикет
df_irkut3 = pd.read_csv("ИРКУТ_3") #третий тикет

In [35]:
#посмотрим как выглядят данные
print("Первые десять строк в таблице:")
display(df_transfap.head())
display(df_krasoctao.head())
display(df_irkut3.head())
display(df_transfap.iloc[-1])

Первые десять строк в таблице:


,<TICKER>,<PER>,<DATE>,<TIME>,<CLOSE>,<VOL>
0,Транснф_ап,D,20081222,0,7536.84,2689
1,Транснф_ап,D,20081223,0,7748.00,6396
2,Транснф_ап,D,20081224,0,7576.93,4259
3,Транснф_ап,D,20081225,0,7350.00,4529
4,Транснф_ап,D,20081226,0,7011.11,2712


,<TICKER>,<PER>,<DATE>,<TIME>,<CLOSE>,<VOL>
0,КрасОкт-ао,D,20081222,0,213.0,142
1,КрасОкт-ао,D,20081223,0,200.0,136
2,КрасОкт-ао,D,20081224,0,190.0,121
3,КрасОкт-ао,D,20081225,0,175.0,420
4,КрасОкт-ао,D,20081226,0,175.0,15


,<TICKER>,<PER>,<DATE>,<TIME>,<CLOSE>,<VOL>
0,ИРКУТ-3,D,20081222,0,3.300,1500
1,ИРКУТ-3,D,20081223,0,3.410,28600
2,ИРКУТ-3,D,20081224,0,3.550,147900
3,ИРКУТ-3,D,20081225,0,3.400,97000
4,ИРКУТ-3,D,20081226,0,3.525,130500


<TICKER>    Транснф_ап
<PER>                D
<DATE>        20220719
<TIME>               0
<CLOSE>       112300.0
<VOL>             2009
Name: 4957, dtype: object

In [36]:
#удалим ненужные столбцы
df_transfap.drop(["<TICKER>", "<PER>", "<TIME>"], axis=1, inplace=True)
df_krasoctao.drop(["<TICKER>", "<PER>", "<TIME>"], axis=1, inplace=True)
df_irkut3.drop(["<TICKER>", "<PER>", "<TIME>"], axis=1, inplace=True)

In [37]:
#поменяем название оставшихся
df_transfap.rename(columns={"<DATE>": "Дата"}, inplace=True)
df_transfap.rename(columns={"<CLOSE>":"Цена"}, inplace=True)
df_transfap.rename(columns={"<VOL>":"Объем"}, inplace=True)

df_krasoctao.rename(columns={"<DATE>": "Дата"}, inplace=True)
df_krasoctao.rename(columns={"<CLOSE>":"Цена"}, inplace=True)
df_krasoctao.rename(columns={"<VOL>":"Объем"}, inplace=True)

df_irkut3.rename(columns={"<DATE>": "Дата"}, inplace=True)
df_irkut3.rename(columns={"<CLOSE>":"Цена"}, inplace=True)
df_irkut3.rename(columns={"<VOL>":"Объем"}, inplace=True)

In [38]:
df_transfap.head()

,Дата,Цена,Объем
0,20081222,7536.84,2689
1,20081223,7748.00,6396
2,20081224,7576.93,4259
3,20081225,7350.00,4529
4,20081226,7011.11,2712


In [39]:
#Поменяем формат столбца Дата
df_transfap["Дата"] = pd.to_datetime(df_transfap["Дата"], format='%Y%m%d')
df_krasoctao["Дата"] = pd.to_datetime(df_krasoctao["Дата"], format = '%Y%m%d')
df_irkut3["Дата"] = pd.to_datetime(df_irkut3["Дата"], format = '%Y%m%d')

### Очистка данных

In [40]:
display(df_transfap.describe())
display(df_krasoctao.describe())
display(df_irkut3.describe())

,Дата,Цена,Объем
count,4958,4958.000000,4958.000000
mean,2015-10-05 12:00:00,110711.829322,5580.297701
min,2008-12-22 00:00:00,6122.960000,0.000000
25%,2012-05-14 06:00:00,53517.000000,0.000000
50%,2015-10-05 12:00:00,131805.000000,1612.500000
75%,2019-02-25 18:00:00,162600.000000,5552.250000
max,2022-07-19 00:00:00,231000.000000,182570.000000
std,NaN,58579.818604,11596.650253


,Дата,Цена,Объем
count,4958,4958.000000,4958.000000
mean,2015-10-05 12:00:00,290.013584,1385.917104
min,2008-12-22 00:00:00,85.000000,0.000000
25%,2012-05-14 06:00:00,180.005000,0.000000
50%,2015-10-05 12:00:00,305.120000,120.000000
75%,2019-02-25 18:00:00,366.000000,769.750000
max,2022-07-19 00:00:00,840.000000,289610.000000
std,NaN,122.603433,9027.927276


,Дата,Цена,Объем
count,4958,4958.000000,4.958000e+03
mean,2015-10-05 12:00:00,15.382586,1.532985e+05
min,2008-12-22 00:00:00,3.100000,0.000000e+00
25%,2012-05-14 06:00:00,6.330000,0.000000e+00
50%,2015-10-05 12:00:00,9.200000,4.230000e+04
75%,2019-02-25 18:00:00,24.320000,1.244250e+05
max,2022-07-19 00:00:00,77.150000,1.833340e+07
std,NaN,12.635164,6.235901e+05


In [41]:
#проверим совпадают ли даты у всех тикеров
are_equal = df_transfap['Дата'].equals(df_krasoctao['Дата']) and df_krasoctao['Дата'].equals(df_irkut3['Дата'])

if are_equal:
    print("Все даты во всех таблицах совпадают идеально.")
else:
    print("Есть расхождения в датах или длине таблиц.")

Все даты во всех таблицах совпадают идеально.


.reset_index(drop=True) - переписывает все индексы заново

In [42]:
#удалим строки, где объем = 0 (или пропущен)
bad_indices = set()

for df in [df_transfap, df_krasoctao, df_irkut3]:
    invalid_rows = df[(df['Цена'] <= 0) | (df['Объем'] <= 0) | 
                      (df['Цена'].isna()) | (df['Объем'].isna())].index
    bad_indices.update(invalid_rows)

df_transfap = df_transfap.drop(index=bad_indices).reset_index(drop=True)
df_krasoctao = df_krasoctao.drop(index=bad_indices).reset_index(drop=True)
df_irkut3 = df_irkut3.drop(index=bad_indices).reset_index(drop=True)

In [43]:
display(df_transfap.describe())
display(df_krasoctao.describe())
display(df_irkut3.describe())

,Дата,Цена,Объем
count,3048,3048.000000,3048.000000
mean,2015-11-14 14:11:20.314960,111503.288583,8588.936352
min,2008-12-22 00:00:00,6289.000000,21.000000
25%,2012-01-26 18:00:00,47887.500000,1459.750000
50%,2016-04-19 00:00:00,136875.000000,3151.500000
75%,2019-06-10 06:00:00,164275.000000,10291.000000
max,2022-07-19 00:00:00,231000.000000,182570.000000
std,NaN,60095.028104,13868.410013


,Дата,Цена,Объем
count,3048,3048.000000,3048.000000
mean,2015-11-14 14:11:20.314960,301.732815,2254.378937
min,2008-12-22 00:00:00,85.000000,1.000000
25%,2012-01-26 18:00:00,200.000000,179.500000
50%,2016-04-19 00:00:00,313.215000,540.000000
75%,2019-06-10 06:00:00,375.552500,1362.500000
max,2022-07-19 00:00:00,840.000000,289610.000000
std,NaN,123.064822,11429.572640


,Дата,Цена,Объем
count,3048,3048.000000,3.048000e+03
mean,2015-11-14 14:11:20.314960,16.420467,2.347178e+05
min,2008-12-22 00:00:00,3.100000,3.000000e+02
25%,2012-01-26 18:00:00,6.713500,4.250000e+04
50%,2016-04-19 00:00:00,9.700000,9.030000e+04
75%,2019-06-10 06:00:00,24.940000,1.966000e+05
max,2022-07-19 00:00:00,77.150000,1.833340e+07
std,NaN,13.021606,7.577929e+05


Было 4958 значений, после удаления стало 3048 значений. 

### Логдоходность

Для расчета дневной логарифмической доходности используем формулу:$$r_t = \ln\left(\frac{P_t}{P_{t-1}}\right)$$

In [44]:
#функция расчета логдоходности

def calc_logprofit(df):

    df["Ln"] = np.log(df["Цена"] / df["Цена"].shift(1))
    
    df = df.iloc[1:].reset_index(drop=True)

    return df

df_transfap = calc_logprofit(df_transfap)
df_krasoctao = calc_logprofit(df_krasoctao)
df_irkut3 = calc_logprofit(df_irkut3)

# Проверим результат 
display(df_transfap.head())
display(df_krasoctao.head())

,Дата,Цена,Объем,Ln
0,2008-12-23,7748.00,6396,0.027632
1,2008-12-24,7576.93,4259,-0.022327
2,2008-12-25,7350.00,4529,-0.030408
3,2008-12-26,7011.11,2712,-0.047204
4,2008-12-29,6830.00,2573,-0.026171


,Дата,Цена,Объем,Ln
0,2008-12-23,200.0,136,-0.062975
1,2008-12-24,190.0,121,-0.051293
2,2008-12-25,175.0,420,-0.082238
3,2008-12-26,175.0,15,0.000000
4,2008-12-29,180.0,32,0.028171


In [45]:
#количество строк, в которых хотя бы для одной компании логдоходность равна нулю:
print(len(df_transfap))
bad_indices = set()

for df in [df_transfap, df_krasoctao, df_irkut3]:
    invalid_rows = df[(df['Ln'] == 0)].index
    bad_indices.update(invalid_rows)

print(len(bad_indices))

df_transfap = df_transfap.drop(index=bad_indices).reset_index(drop=True)
df_krasoctao = df_krasoctao.drop(index=bad_indices).reset_index(drop=True)
df_irkut3 = df_irkut3.drop(index=bad_indices).reset_index(drop=True)

print(len(df_transfap))

3047
319
2728


### Анализ Данных

In [46]:
#выборочное среднее значение логдоходности
display(df_transfap.describe())
display(df_krasoctao.describe())
display(df_irkut3.describe())


,Дата,Цена,Объем,Ln
count,2728,2728.000000,2728.000000,2728.000000
mean,2015-09-27 17:04:34.486803,109132.981910,8979.373167,0.000992
min,2008-12-23 00:00:00,6318.000000,21.000000,-0.325370
25%,2011-12-01 18:00:00,44586.250000,1513.750000,-0.010095
50%,2016-01-25 12:00:00,133980.000000,3469.500000,0.000379
75%,2019-05-14 12:00:00,163000.000000,11074.250000,0.010635
max,2022-07-19 00:00:00,231000.000000,182570.000000,0.199636
std,NaN,60369.899527,14115.709682,0.025425


,Дата,Цена,Объем,Ln
count,2728,2728.000000,2728.000000,2728.000000
mean,2015-09-27 17:04:34.486803,300.679575,2294.467742,-0.000044
min,2008-12-23 00:00:00,85.000000,1.000000,-0.509009
25%,2011-12-01 18:00:00,200.000000,190.000000,-0.015748
50%,2016-01-25 12:00:00,312.720000,560.000000,-0.002445
75%,2019-05-14 12:00:00,373.000000,1410.000000,0.012749
max,2022-07-19 00:00:00,840.000000,289610.000000,0.479729
std,NaN,122.314726,11591.767290,0.046526


,Дата,Цена,Объем,Ln
count,2728,2728.000000,2.728000e+03,2728.000000
mean,2015-09-27 17:04:34.486803,16.156598,2.419663e+05,0.000754
min,2008-12-23 00:00:00,3.100000,3.000000e+02,-0.322623
25%,2011-12-01 18:00:00,6.679750,4.317500e+04,-0.010993
50%,2016-01-25 12:00:00,9.480000,9.150000e+04,-0.001322
75%,2019-05-14 12:00:00,24.765000,2.008500e+05,0.010191
max,2022-07-19 00:00:00,77.150000,1.833340e+07,0.262364
std,NaN,12.913137,7.901671e+05,0.031882


In [47]:
#корреляция между столбцами Ln
corr_trans_kras = df_transfap['Ln'].corr(df_krasoctao['Ln'])
corr_trans_irkut = df_transfap['Ln'].corr(df_irkut3['Ln'])
corr_kras_irkut = df_krasoctao['Ln'].corr(df_irkut3['Ln'])

print(f"ρ(Транснф ап, КрасОкт-ао) = {corr_trans_kras:.12f}")
print(f"ρ(Транснф ап, ИРКУТ-3) = {corr_trans_irkut:.12f}")
print(f"ρ(КрасОкт-ао, ИРКУТ-3) = {corr_kras_irkut:.12f}")

ρ(Транснф ап, КрасОкт-ао) = 0.085477573094
ρ(Транснф ап, ИРКУТ-3) = 0.207393278009
ρ(КрасОкт-ао, ИРКУТ-3) = 0.139859857765


In [48]:
print(f"Среднее КрасОкт-ао: {df_krasoctao['Ln'].mean():.12f}")
print(f"Среднее ИРКУТ-3:    {df_irkut3['Ln'].mean():.12f}")

Среднее КрасОкт-ао: -0.000043775269
Среднее ИРКУТ-3:    0.000753523048


In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

mean_trans = df_transfap['Ln'].mean()
mean_kras = df_krasoctao['Ln'].mean()
mean_irkut = df_irkut3['Ln'].mean()

std_trans = df_transfap['Ln'].std(ddof=1)
std_kras = df_krasoctao['Ln'].std(ddof=1)
std_irkut = df_irkut3['Ln'].std(ddof=1)

# 2. Теперь собираем данные для диаграммы Парето
assets = [
    {'name': 'Транснф ап', 'mean': mean_trans, 'std': std_trans},
    {'name': 'КрасОкт-ао', 'mean': mean_kras, 'std': std_kras},
    {'name': 'ИРКУТ-3', 'mean': mean_irkut, 'std': std_irkut}
]

# 3. Настройка графиков
sns.set_theme(style="whitegrid")
fig = plt.figure(figsize=(15, 12))

# --- 2.1 Диаграммы рассеяния ---
ax1 = fig.add_subplot(3, 2, 1)
sns.scatterplot(x=df_transfap['Ln'], y=df_krasoctao['Ln'], ax=ax1, alpha=0.5)
ax1.set_title('Рассеяние: Транснф ап vs КрасОкт-ао')

ax2 = fig.add_subplot(3, 2, 2)
sns.scatterplot(x=df_transfap['Ln'], y=df_irkut3['Ln'], ax=ax2, color='orange', alpha=0.5)
ax2.set_title('Рассеяние: Транснф ап vs ИРКУТ-3')

ax3 = fig.add_subplot(3, 2, 3)
sns.scatterplot(x=df_krasoctao['Ln'], y=df_irkut3['Ln'], ax=ax3, color='green', alpha=0.5)
ax3.set_title('Рассеяние: КрасОкт-ао vs ИРКУТ-3')

# --- 2.2 Диаграмма Парето ---
ax4 = fig.add_subplot(3, 2, 4)
stds = [a['std'] for a in assets]
means = [a['mean'] for a in assets]
ax4.scatter(stds, means, color='red', s=100)
for a in assets:
    ax4.annotate(a['name'], (a['std'], a['mean']), xytext=(5,5), textcoords='offset points')
ax4.set_title('Диаграмма Парето')
ax4.set_xlabel('Риск (Std Dev)')
ax4.set_ylabel('Доходность (Mean)')

# --- 2.3 Диаграммы размаха (Boxplots) ---
ax5 = fig.add_subplot(3, 1, 3)
box_data = pd.DataFrame({
    'Транснф ап': df_transfap['Ln'],
    'КрасОкт-ао': df_krasoctao['Ln'],
    'ИРКУТ-3': df_irkut3['Ln']
})
sns.boxplot(data=box_data, orient='h', ax=ax5, palette="Set2")
ax5.set_title('Диаграммы размаха логдоходностей')

plt.tight_layout()
plt.show()

In [ ]:


# 1. Считаем финальные показатели (убедись, что даты уже обрезаны по 19.07.2022)
data = {
    'Ticker': ['Транснф ап', 'КрасОкт-ао', 'ИРКУТ-3'],
    'Mean': [df_transfap['Ln'].mean(), df_krasoctao['Ln'].mean(), df_irkut3['Ln'].mean()],
    'Std': [df_transfap['Ln'].std(), df_krasoctao['Ln'].std(), df_irkut3['Ln'].std()]
}
res = pd.DataFrame(data)

# 2. Логика определения оптимальности (Парето)
optimal_flags = []

for i, row in res.iterrows():
    is_optimal = True
    for j, other in res.iterrows():
        if i == j: continue
        
        # Проверяем, есть ли кто-то ЛУЧШЕ по обоим параметрам сразу
        # (Риск меньше или равен И Доходность больше или равна)
        better_risk = other['Std'] <= row['Std']
        better_return = other['Mean'] >= row['Mean']
        
        # Хотя бы один параметр должен быть строго лучше, чтобы "выбить" акцию из Парето
        strictly_better = (other['Std'] < row['Std']) or (other['Mean'] > row['Mean'])
        
        if better_risk and better_return and strictly_better:
            is_optimal = False
            break
    optimal_flags.append(1 if is_optimal else 0)

res['Optimal'] = optimal_flags

# 3. Формируем двоичный код для ответа (Часть 12)
binary_code = "".join(map(str, optimal_flags))
# Убираем ведущий ноль, если он есть (например, 011 -> 11)
final_answer = str(int(binary_code))

print(f"Результаты расчетов:\n{res}\n")
print(f"ОТВЕТ ЗА ЧАСТЬ 12 (Двоичный код): {final_answer}")

# 4. Построение ПРАВИЛЬНОЙ диаграммы
plt.figure(figsize=(10, 6))
for i, row in res.iterrows():
    color = 'green' if row['Optimal'] == 1 else 'red'
    plt.scatter(row['Std'], row['Mean'], c=color, s=100, edgecolors='black')
    plt.annotate(row['Ticker'], (row['Std'], row['Mean']), xytext=(5, 5), textcoords='offset points')

plt.title('Диаграмма Парето (Зеленые - оптимальные, Красные - нет)')
plt.xlabel('Риск (Std)')
plt.ylabel('Доходность (Mean)')
plt.grid(True, linestyle='--', alpha=0.6)
plt.show()

In [ ]:
# 1. Проверка длин таблиц (должны совпадать)
len_trans = len(df_transfap)
len_kras = len(df_krasoctao)
len_irkut = len(df_irkut3)

print("--- 1. ПРОВЕРКА СИНХРОННОСТИ ---")
if len_trans == len_kras == len_irkut:
    print(f"✅ ОК: Все таблицы имеют одинаковый размер ({len_trans} строк).")
else:
    print(f"❌ ОШИБКА: Размеры не совпадают! Транснф: {len_trans}, КрасОкт: {len_kras}, ИРКУТ: {len_irkut}")

# 2. Проверка дат
print("\n--- 2. ПРОВЕРКА ДАТ ---")
min_date = df_transfap['Дата'].min()
max_date = df_transfap['Дата'].max()
print(f"Начало данных: {min_date}")
print(f"Конец данных: {max_date}")
if str(min_date)[:10] < '2008-12-22' or str(max_date)[:10] > '2022-07-19':
    print("❌ ОШИБКА: Даты выходят за рамки задания (2008-12-22 - 2022-07-19)!")
else:
    print("✅ ОК: Даты в пределах заданного интервала.")

# 3. Проверка на "битые" значения (NaN или бесконечности)
print("\n--- 3. ПРОВЕРКА ЛОГДОХОДНОСТИ ---")
has_na = df_transfap['Ln'].isna().sum() + df_krasoctao['Ln'].isna().sum() + df_irkut3['Ln'].isna().sum()
if has_na > 0:
    print(f"❌ ОШИБКА: В столбцах Ln найдено {has_na} пустых значений (NaN)!")
else:
    print("✅ ОК: Пустых значений в логдоходности нет.")